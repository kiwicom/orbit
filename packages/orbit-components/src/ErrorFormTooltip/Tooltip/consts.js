// @flow

export const SIDE_NUDGE = 15;
