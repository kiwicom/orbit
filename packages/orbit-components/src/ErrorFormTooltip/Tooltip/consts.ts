export const SIDE_NUDGE = 10;
