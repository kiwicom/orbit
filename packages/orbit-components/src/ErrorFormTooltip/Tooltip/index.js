// @flow
import * as React from "react";
import styled, { css } from "styled-components";

import useClickOutside from "../../hooks/useClickOutside";
import KEY_CODE_MAP from "../../common/keyMaps";
import handleKeyDown from "../../utils/handleKeyDown";
import defaultTheme from "../../defaultTheme";
import media from "../../utils/mediaQuery";
import { StyledText } from "../../Text";
import { Item } from "../../List/ListItem";
import CloseIc from "../../icons/Close";
import { rtlSpacing, right } from "../../utils/rtl";
import resolveColor from "./helpers/resolveColor";
import tooltipArrowStyle from "./helpers/tooltipArrowStyle";
import resolveTooltipArrowPosition from "./helpers/resolveTooltipArrowPosition";
import resolveTooltipPosition from "./helpers/resolveTooltipPosition";
import useDimensions from "../hooks/useDimensions";
import { POSITIONS, SIDE_NUDGE } from "./consts";

import type { Props } from ".";

const StyledFormFeedbackTooltip = styled.div`
  ${({ theme, isHelp, shown, inputSize }) => css`
    display: flex;
    justify-content: space-between;
    align-items: center;
    position: absolute;
    box-sizing: border-box;
    border-radius: ${theme.orbit.borderRadiusNormal};
    box-shadow: ${theme.orbit.boxShadowElevatedLevel1};
    padding: ${inputSize === "small" ? theme.orbit.spaceXSmall : theme.orbit.spaceSmall};
    padding-${right}: ${isHelp && theme.orbit.spaceSmall};
    z-index: 10;
    max-height: none;
    overflow: visible;
    width: ${`calc(100% + ${SIDE_NUDGE * 2}px)`};
    background-color: ${resolveColor};
    visibility: ${shown ? "visible" : "hidden"};
    opacity: ${shown ? "1" : "0"};
    transition: opacity ${theme.orbit.durationFast} ease-in-out,
      visibility ${theme.orbit.durationFast} ease-in-out;

    /* prevent position, IEs don't have initial YAY */
    top: auto;
    right: auto;
    bottom: auto;
    left: auto;

    img {
      max-width: 100%;
    }

    &::after {
      width: 0;
      height: 0;
      border-style: solid;
      content: " ";
      display: block;
      position: absolute;

      ${tooltipArrowStyle};
      ${resolveTooltipArrowPosition};
    }

    ${media.largeMobile(css`
      width: auto;
    `)};

    ${resolveTooltipPosition};
`}
`;

// $FlowFixMe: https://github.com/flow-typed/flow-typed/issues/3653#issuecomment-568539198
StyledFormFeedbackTooltip.defaultProps = {
  theme: defaultTheme,
};

const StyledTooltipContent = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    height: 100%;
    font-family: ${theme.orbit.fontFamily};
    font-size: ${theme.orbit.fontSizeTextNormal};
    font-weight: ${theme.orbit.fontWeightNormal};
    line-height: ${theme.orbit.lineHeightText};
    color: ${theme.orbit.paletteWhite};

    & ${StyledText}, ${Item}, a {
      color: ${theme.orbit.paletteWhite};
      font-weight: ${theme.orbit.fontWeightNormal};
    }

    ${media.largeMobile(css`
      font-size: ${theme.orbit.fontSizeTextSmall};
      font-weight: ${theme.orbit.fontWeightMedium};

      ${StyledText}, ${Item}, a {
        font-weight: ${theme.orbit.fontWeightMedium};
        font-size: ${theme.orbit.fontSizeTextSmall};
      }
      a:hover {
        color: ${theme.orbit.paletteWhite};
      }
      a:focus {
        color: ${theme.orbit.paletteWhite};
      }
    `)};
  `}
`;

// $FlowFixMe: https://github.com/flow-typed/flow-typed/issues/3653#issuecomment-568539198
StyledTooltipContent.defaultProps = {
  theme: defaultTheme,
};

const StyledCloseButton = styled.a`
  ${({ theme }) => css`
    color: ${theme.orbit.paletteWhite};
    cursor: pointer;
    margin: ${rtlSpacing(`0 0 0 ${theme.orbit.spaceSmall}`)};
    display: flex;
  `}
`;

// $FlowFixMe: https://github.com/flow-typed/flow-typed/issues/3653#issuecomment-568539198
StyledCloseButton.defaultProps = {
  theme: defaultTheme,
};

const ErrorFormTooltip = ({
  onShown,
  labelRef,
  dataTest,
  inputSize,
  iconRef,
  children,
  shown,
  isHelp = false,
  inlineLabel,
  id,
}: Props): React.Node => {
  const contentRef = React.useRef(null);
  const tooltipRef = React.useRef<HTMLElement | null>(null);

  const dimensions = useDimensions({
    labelRef,
    iconRef,
    contentRef: tooltipRef,
    children,
    inlineLabel,
  });

  const preferredPosition = React.useMemo(
    () => (dimensions.label.top - dimensions.content.height > 0 ? POSITIONS.TOP : POSITIONS.BOTTOM),
    [dimensions.label.top, dimensions.content.height],
  );

  useClickOutside(contentRef, () => {
    if (!isHelp) onShown(false);
  });

  React.useEffect(() => {
    // $FlowFixMe: TODO
    const link = tooltipRef.current?.querySelector("a");

    const handleTab = ev => {
      if (isHelp) return;
      if (ev.keyCode === KEY_CODE_MAP.TAB && link) {
        onShown(true);
        if (document.activeElement === link) {
          onShown(false);
        }
      } else {
        onShown(false);
      }
    };

    window.addEventListener("keydown", handleTab);
    return () => {
      window.removeEventListener("keydown", handleTab);
    };
  }, [onShown, isHelp]);

  return (
    <StyledFormFeedbackTooltip
      ref={tooltipRef}
      inputSize={inputSize}
      position={preferredPosition}
      shown={shown && dimensions.set}
      isHelp={isHelp}
      data-test={dataTest}
      inlineLabel={inlineLabel}
      aria-live="polite"
      id={id}
      {...dimensions}
    >
      <StyledTooltipContent ref={contentRef}>{children}</StyledTooltipContent>
      {isHelp && (
        <StyledCloseButton
          tabIndex={0}
          // $FlowFixMe: TODO
          onKeyDown={handleKeyDown(onShown)}
          onClick={ev => {
            ev.preventDefault();
            if (shown) onShown(false);
          }}
        >
          <CloseIc ariaLabel="close" size="small" />
        </StyledCloseButton>
      )}
    </StyledFormFeedbackTooltip>
  );
};

export default ErrorFormTooltip;
