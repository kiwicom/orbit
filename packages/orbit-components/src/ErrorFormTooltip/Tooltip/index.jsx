// @flow
import * as React from "react";
import styled, { css } from "styled-components";
import { usePopper } from "react-popper";

import useClickOutside from "../../hooks/useClickOutside";
import KEY_CODE_MAP from "../../common/keyMaps";
import handleKeyDown from "../../utils/handleKeyDown";
import defaultTheme from "../../defaultTheme";
import media from "../../utils/mediaQuery";
import { StyledText } from "../../Text";
import { Item } from "../../List/ListItem";
import CloseIc from "../../icons/Close";
import { rtlSpacing, right } from "../../utils/rtl";
import resolveColor from "./helpers/resolveColor";
import resolvePlacement from "./helpers/resolvePlacement";
import { SIDE_NUDGE } from "./consts";
import useTheme from "../../hooks/useTheme";

import type { Props } from ".";

const StyledArrow = styled.div`
  ${({ position, top, left, bottom, right: popperRight, transform }) => css`
    position: ${position};
    top: ${top};
    left: ${left};
    right: ${popperRight};
    bottom: ${bottom};
    transform: ${transform};
    &:before {
      content: "";
      background: ${resolveColor};
      width: 0.6rem;
      height: 0.6rem;
      transform: translate(-50%, -50%) rotate(45deg);
      position: absolute;
    }
  `}
`;

const StyledFormFeedbackTooltip = styled.div`
  ${({
    theme,
    isHelp,
    shown,
    inputSize,
    top,
    left,
    position,
    bottom,
    right: popperRight,
    transform,
  }) => css`
    display: flex;
    justify-content: space-between;
    box-sizing: border-box;
    border-radius: ${theme.orbit.borderRadiusNormal};
    box-shadow: ${theme.orbit.boxShadowElevatedLevel1};
    padding: ${theme.orbit.spaceXSmall} ${
    inputSize === "small" ? theme.orbit.spaceXSmall : theme.orbit.spaceSmall
  };
    padding-${right}: ${isHelp && theme.orbit.spaceSmall};
    z-index: 10;
    max-height: none;
    overflow: visible;
    width: ${`calc(100% + ${SIDE_NUDGE * 2}px)`};
    width: min(${`calc(100% + ${SIDE_NUDGE * 2}px)`}, 100vw);
    background: ${resolveColor};
    visibility: ${shown ? "visible" : "hidden"};
    opacity: ${shown ? "1" : "0"};
    transition: opacity ${theme.orbit.durationFast} ease-in-out,
      visibility ${theme.orbit.durationFast} ease-in-out;

    position: ${position};
    top: ${top};
    left: ${left};
    bottom: ${bottom};
    right: ${popperRight};
    transform: ${transform};

    img {
      max-width: 100%;
    }

    ${StyledArrow} {
      ${resolvePlacement};
    }

    ${media.largeMobile(css`
      width: auto;
    `)};
`}
`;

// $FlowFixMe: https://github.com/flow-typed/flow-typed/issues/3653#issuecomment-568539198
StyledArrow.defaultProps = {
  theme: defaultTheme,
};

// $FlowFixMe: https://github.com/flow-typed/flow-typed/issues/3653#issuecomment-568539198
StyledFormFeedbackTooltip.defaultProps = {
  theme: defaultTheme,
};

const StyledTooltipContent = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    height: 100%;
    font-family: ${theme.orbit.fontFamily};
    font-size: ${theme.orbit.fontSizeTextNormal};
    font-weight: ${theme.orbit.fontWeightNormal};
    line-height: ${theme.orbit.lineHeightText};
    color: ${theme.orbit.paletteWhite};

    & ${StyledText}, ${Item}, a {
      color: ${theme.orbit.paletteWhite};
      font-size: ${theme.orbit.fontSizeTextNormal};
      font-weight: ${theme.orbit.fontWeightNormal};
      &:hover,
      &:focus {
        color: ${theme.orbit.paletteWhite};
      }
    }

    ${media.largeMobile(css`
      font-size: ${theme.orbit.fontSizeTextSmall};
      font-weight: ${theme.orbit.fontWeightMedium};

      ${StyledText}, ${Item}, a {
        font-weight: ${theme.orbit.fontWeightMedium};
        font-size: ${theme.orbit.fontSizeTextSmall};
      }
    `)};
  `}
`;

// $FlowFixMe: https://github.com/flow-typed/flow-typed/issues/3653#issuecomment-568539198
StyledTooltipContent.defaultProps = {
  theme: defaultTheme,
};

const StyledCloseButton = styled.a`
  ${({ theme }) => css`
    color: ${theme.orbit.paletteWhite};
    cursor: pointer;
    margin: ${rtlSpacing(`0 0 0 ${theme.orbit.spaceSmall}`)};
    display: flex;
  `}
`;

// $FlowFixMe: https://github.com/flow-typed/flow-typed/issues/3653#issuecomment-568539198
StyledCloseButton.defaultProps = {
  theme: defaultTheme,
};

const ErrorFormTooltip = ({
  onShown,
  dataTest,
  helpClosable,
  inputSize,
  children,
  shown,
  referenceElement,
  inlineLabel,
  isHelp = false,
  id,
}: Props): React.Node => {
  const contentRef = React.useRef(null);
  const { rtl } = useTheme();
  const tooltipRef = React.useRef(null);
  const [arrowRef, setArrowRef] = React.useState(null);

  const resolveOffset = React.useCallback(() => {
    if (inlineLabel) {
      if (inputSize === "small") return [rtl ? 10 : -14, 7];
      return [rtl ? 6 : -6, 6];
    }
    return [rtl ? SIDE_NUDGE : -SIDE_NUDGE, 7];
  }, [inlineLabel, inputSize, rtl]);

  const { styles, attributes: attrs, update } = usePopper(
    referenceElement?.current,
    tooltipRef.current,
    {
      placement: rtl ? "top-end" : "top-start",
      modifiers: [
        {
          name: "offset",
          options: {
            offset: resolveOffset,
          },
        },
        {
          name: "arrow",
          options: {
            element: arrowRef,
          },
        },
        {
          name: "eventListeners",
          options: {
            scroll: false,
          },
        },
      ],
    },
  );

  const { popper, arrow } = styles;

  useClickOutside(tooltipRef, () => {
    onShown(false);
  });

  React.useEffect(() => {
    if (update) update();
  }, [update, resolveOffset, shown]);

  React.useEffect(() => {
    const link = tooltipRef.current?.querySelector("a");
    const handleTab = ev => {
      if (isHelp) return;
      if (ev.keyCode === KEY_CODE_MAP.TAB && link) {
        onShown(true);
        if (document.activeElement === link) {
          onShown(false);
        }
      } else {
        onShown(false);
      }
    };

    window.addEventListener("keydown", handleTab);
    return () => {
      window.removeEventListener("keydown", handleTab);
    };
  }, [onShown, isHelp, helpClosable]);

  return (
    <StyledFormFeedbackTooltip
      id={id}
      ref={tooltipRef}
      inputSize={inputSize}
      shown={shown}
      isHelp={isHelp}
      data-test={dataTest}
      aria-live="polite"
      inlineLabel={inlineLabel}
      placement={attrs.popper && attrs.popper["data-popper-placement"]}
      position={popper.position}
      top={popper.top}
      left={popper.left}
      right={popper.right}
      bottom={popper.bottom}
      transform={popper.transform}
    >
      <StyledArrow
        isHelp={isHelp}
        ref={setArrowRef}
        position={popper.position}
        placement={attrs.popper && attrs.popper["data-popper-placement"]}
        top={arrow.top}
        left={arrow.left}
        right={arrow.right}
        bottom={arrow.bottom}
      />
      <StyledTooltipContent ref={contentRef}>{children}</StyledTooltipContent>
      {isHelp && helpClosable && (
        <StyledCloseButton
          tabIndex={0}
          // $FlowFixMe: TODO
          onKeyDown={handleKeyDown(onShown)}
          onClick={ev => {
            ev.preventDefault();
            if (shown) onShown(false);
          }}
        >
          <CloseIc ariaLabel="close" size="small" />
        </StyledCloseButton>
      )}
    </StyledFormFeedbackTooltip>
  );
};

export default ErrorFormTooltip;
