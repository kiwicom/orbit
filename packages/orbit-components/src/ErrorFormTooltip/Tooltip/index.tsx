import * as React from "react";
import styled, { css } from "styled-components";
import { usePopper } from "react-popper";

import * as Common from "../../common/types";
import useClickOutside from "../../hooks/useClickOutside";
import KEY_CODE_MAP from "../../common/keyMaps";
import handleKeyDown from "../../utils/handleKeyDown";
import defaultTheme from "../../defaultTheme";
import media from "../../utils/mediaQuery";
import { StyledText } from "../../Text";
import { Item } from "../../List/ListItem";
import CloseIc from "../../icons/Close";
import { rtlSpacing, right } from "../../utils/rtl";
import resolveColor from "./helpers/resolveColor";
import resolvePlacement from "./helpers/resolvePlacement";
import { SIDE_NUDGE } from "./consts";
import useTheme from "../../hooks/useTheme";
import { Props } from "./types";

const StyledArrow = styled.div<{
  inlineLabel?: boolean;
  isHelp?: boolean;
  inputSize?: Common.InputSize;
  placement?: string;
}>`
  position: absolute;
  ${resolvePlacement};
  &:before {
    content: "";
    background: ${resolveColor};
    width: 8px;
    height: 8px;
    transform: translate(-50%, -50%) rotate(45deg);
    position: absolute;
  }
`;

const StyledFormFeedbackTooltip = styled.div<{
  isHelp?: boolean;
  shown?: boolean;
  inlineLabel?: boolean;
  inputSize?: Common.InputSize;
  placement?: string;
  top?: string | number;
  left?: string | number;
  position?: string | number;
  bottom?: string | number;
  right?: string | number;
  transform?: string;
}>`
  ${({
    theme,
    isHelp,
    shown,
    inputSize,
    top,
    left,
    position,
    bottom,
    right: popperRight,
    transform,
  }) => css`
    display: flex;
    justify-content: space-between;
    box-sizing: border-box;
    border-radius: ${theme.orbit.borderRadiusNormal};
    box-shadow: ${theme.orbit.boxShadowElevatedLevel1};
    padding: ${theme.orbit.spaceXSmall} ${
    inputSize === "small" ? theme.orbit.spaceXSmall : theme.orbit.spaceSmall
  };
    padding-${right}: ${isHelp && theme.orbit.spaceSmall};
    z-index: 10;
    max-height: none;
    overflow: visible;
    width: min(${`calc(100% - ${SIDE_NUDGE * 2}px)`}, 100vw);
    background: ${resolveColor};
    visibility: ${shown ? "visible" : "hidden"};
    opacity: ${shown ? "1" : "0"};
    transition: opacity ${theme.orbit.durationFast} ease-in-out,
      visibility ${theme.orbit.durationFast} ease-in-out;

    position: ${position};
    top: ${top};
    left: ${left};
    bottom: ${bottom};
    right: ${popperRight};
    transform: ${transform};

    img {
      max-width: 100%;
    }

    ${media.largeMobile(css`
      width: auto;
    `)};
`}
`;

StyledArrow.defaultProps = {
  theme: defaultTheme,
};

StyledFormFeedbackTooltip.defaultProps = {
  theme: defaultTheme,
};

const StyledTooltipContent = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    height: 100%;
    font-family: ${theme.orbit.fontFamily};
    font-size: ${theme.orbit.fontSizeTextNormal};
    font-weight: ${theme.orbit.fontWeightNormal};
    line-height: ${theme.orbit.lineHeightText};
    color: ${theme.orbit.paletteWhite};

    & ${StyledText}, ${Item}, a {
      color: ${theme.orbit.paletteWhite};
      font-size: ${theme.orbit.fontSizeTextNormal};
      font-weight: ${theme.orbit.fontWeightNormal};
      &:hover,
      &:focus {
        color: ${theme.orbit.paletteWhite};
      }
    }

    ${media.largeMobile(css`
      font-size: ${theme.orbit.fontSizeTextSmall};
      font-weight: ${theme.orbit.fontWeightMedium};

      ${StyledText}, ${Item}, a {
        font-weight: ${theme.orbit.fontWeightMedium};
        font-size: ${theme.orbit.fontSizeTextSmall};
      }
    `)};
  `}
`;

StyledTooltipContent.defaultProps = {
  theme: defaultTheme,
};

const StyledCloseButton = styled.a`
  ${({ theme }) => css`
    color: ${theme.orbit.paletteWhite};
    cursor: pointer;
    margin: ${rtlSpacing(`0 0 0 ${theme.orbit.spaceSmall}`)};
    display: flex;
  `}
`;

StyledCloseButton.defaultProps = {
  theme: defaultTheme,
};

const ErrorFormTooltip = ({
  onShown,
  dataTest,
  helpClosable,
  inputSize,
  children,
  shown,
  referenceElement,
  inlineLabel,
  isHelp = false,
  id,
}: Props) => {
  const contentRef = React.useRef<HTMLDivElement | null>(null);
  const tooltipRef = React.useRef<HTMLDivElement | null>(null);
  const [arrowRef, setArrowRef] = React.useState<HTMLDivElement | null>(null);
  const { rtl } = useTheme();

  const resolveOffset = React.useCallback(() => {
    if (inlineLabel) {
      if (inputSize === "small") return [0, 4];
    }
    return [0, 3];
  }, [inlineLabel, inputSize]);

  const { styles, attributes: attrs, update } = usePopper(
    referenceElement?.current,
    tooltipRef.current,
    {
      placement: rtl ? "top-end" : "top-start",
      modifiers: [
        {
          name: "offset",
          options: {
            // @ts-expect-error TODO
            offset: resolveOffset,
          },
        },
        {
          name: "flip",
          enabled: false,
        },
        {
          name: "arrow",
          options: {
            element: arrowRef,
          },
        },
        {
          name: "eventListeners",
          options: {
            scroll: false,
          },
        },
      ],
    },
  );

  const { popper } = styles;

  useClickOutside(tooltipRef, () => {
    onShown(false);
  });

  React.useEffect(() => {
    if (update) update();
  }, [update, resolveOffset, shown]);

  React.useEffect(() => {
    const link = tooltipRef.current?.querySelector("a");
    const handleTab = ev => {
      if (isHelp) return;
      if (ev.keyCode === KEY_CODE_MAP.TAB && link) {
        onShown(true);
        if (document.activeElement === link) {
          onShown(false);
        }
      } else {
        onShown(false);
      }
    };

    window.addEventListener("keydown", handleTab);
    return () => {
      window.removeEventListener("keydown", handleTab);
    };
  }, [onShown, isHelp, helpClosable]);

  return (
    <StyledFormFeedbackTooltip
      id={id}
      ref={tooltipRef}
      inputSize={inputSize}
      shown={shown}
      isHelp={isHelp}
      data-test={dataTest}
      aria-live="polite"
      inlineLabel={inlineLabel}
      placement={attrs.popper && attrs.popper["data-popper-placement"]}
      position={popper.position}
      top={popper.top}
      left={popper.left}
      right={popper.right}
      bottom={popper.bottom}
      transform={popper.transform}
    >
      <StyledArrow
        isHelp={isHelp}
        ref={setArrowRef}
        inputSize={inputSize}
        inlineLabel={inlineLabel}
        placement={attrs.popper && attrs.popper["data-popper-placement"]}
      />
      <StyledTooltipContent ref={contentRef}>{children}</StyledTooltipContent>
      {isHelp && helpClosable && (
        <StyledCloseButton
          tabIndex={0}
          // @ts-expect-error TODO
          onKeyDown={handleKeyDown<HTMLAnchorElement>(onShown)}
          onClick={ev => {
            ev.preventDefault();
            if (shown) onShown(false);
          }}
        >
          <CloseIc ariaLabel="close" size="small" />
        </StyledCloseButton>
      )}
    </StyledFormFeedbackTooltip>
  );
};

export default ErrorFormTooltip;
