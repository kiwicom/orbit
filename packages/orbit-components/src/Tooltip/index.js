// @flow
import * as React from "react";

import { SIZE_OPTIONS } from "./consts";
import TooltipPrimitive from "../primitives/TooltipPrimitive";
import useMediaQuery from "../hooks/useMediaQuery";
import MobileDialog from "../primitives/MobileDialogPrimitive";

import type { Props } from "./index";

const Tooltip = ({
  children,
  enabled = true,
  tabIndex = "0",
  dataTest,
  size = SIZE_OPTIONS.SMALL,
  content,
  preferredPosition,
  preferredAlign,
  stopPropagation = false,
  removeUnderlinedText,
}: Props) => {
  const { isLargeMobile } = useMediaQuery();
  return isLargeMobile ? (
    <TooltipPrimitive
      dataTest={dataTest}
      tabIndex={tabIndex}
      enabled={enabled}
      content={content}
      size={size}
      preferredPosition={preferredPosition}
      preferredAlign={preferredAlign}
      stopPropagation={stopPropagation}
      removeUnderlinedText={removeUnderlinedText}
    >
      {children}
    </TooltipPrimitive>
  ) : (
    <MobileDialog
      dataTest={dataTest}
      tabIndex={tabIndex}
      enabled={enabled}
      content={content}
      removeUnderlinedText={removeUnderlinedText}
      stopPropagation={stopPropagation}
    >
      {children}
    </MobileDialog>
  );
};

export default Tooltip;
