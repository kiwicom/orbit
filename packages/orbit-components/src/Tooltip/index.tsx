import * as React from "react";

import { SIZE_OPTIONS } from "./consts";
import TooltipPrimitive from "../primitives/TooltipPrimitive";
import useMediaQuery from "../hooks/useMediaQuery";
import MobileDialog from "../primitives/MobileDialogPrimitive";
import { Props } from "./types";

const Tooltip = ({
  children,
  enabled = true,
  tabIndex = "0",
  dataTest,
  size = SIZE_OPTIONS.SMALL,
  content,
  id,
  placement,
  lockScrolling,
  renderInPortal = true,
  stopPropagation = false,
  removeUnderlinedText,
  block = false,
}: Props) => {
  const { isLargeMobile } = useMediaQuery();
  return isLargeMobile ? (
    <TooltipPrimitive
      dataTest={dataTest}
      id={id}
      tabIndex={tabIndex}
      enabled={enabled}
      content={content}
      size={size}
      renderInPortal={renderInPortal}
      placement={placement}
      stopPropagation={stopPropagation}
      removeUnderlinedText={removeUnderlinedText}
      block={block}
    >
      {children}
    </TooltipPrimitive>
  ) : (
    <MobileDialog
      dataTest={dataTest}
      tabIndex={tabIndex}
      id={id}
      enabled={enabled}
      lockScrolling={lockScrolling}
      content={content}
      removeUnderlinedText={removeUnderlinedText}
      stopPropagation={stopPropagation}
      block={block}
    >
      {children}
    </MobileDialog>
  );
};

export default Tooltip;
