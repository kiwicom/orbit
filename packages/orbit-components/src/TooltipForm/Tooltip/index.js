// @flow
import * as React from "react";
import styled, { css } from "styled-components";

// import handleKeyDown from "../../utils/handleKeyDown";
import defaultTheme from "../../defaultTheme";
import media from "../../utils/mediaQuery";
import { StyledText } from "../../Text";
import { Item } from "../../List/ListItem";
// import CloseIc from "../../icons/Close";
import { rtlSpacing, right } from "../../utils/rtl";
import resolveColor from "./helpers/resolveColor";
import tooltipArrowStyle from "./helpers/tooltipArrowStyle";
import resolveTooltipArrowPosition from "./helpers/resolveTooltipArrowPosition";
import resolveTooltipPosition from "./helpers/resolveTooltipPosition";
import useDimensions from "../hooks/useDimensions";
import { POSITIONS, SIDE_NUDGE, SIZE_OPTIONS } from "./consts";

import type { Props } from "./index";

const getToken = ({ theme, inputSize }) => {
  const tokens = {
    [SIZE_OPTIONS.SMALL]: theme.orbit.heightInputSmall,
    [SIZE_OPTIONS.NORMAL]: theme.orbit.heightInputNormal,
  };

  return tokens[inputSize];
};

const StyledFormFeedbackTooltip = styled.div`
  ${({ theme, isHelp, shown }) => css`
    display: flex;
    justify-content: space-between;
    align-items: center;
    position: absolute;
    box-sizing: border-box;
    border-radius: ${theme.orbit.borderRadiusNormal};
    box-shadow: ${theme.orbit.boxShadowElevatedLevel1};
    padding: 0 ${theme.orbit.spaceSmall};
    padding-${right}: ${isHelp && theme.orbit.spaceSmall};
    height: ${getToken};
    z-index: 10;
    max-height: none;
    overflow: visible;
    width: ${`calc(100% + ${SIDE_NUDGE * 2}px)`};
    background-color: ${resolveColor};
    visibility: ${shown ? "visible" : "hidden"};
    opacity: ${shown ? "1" : "0"};
    transition: opacity ${theme.orbit.durationFast} ease-in-out,
      visibility ${theme.orbit.durationFast} ease-in-out;

    /* prevent position, IEs don't have initial YAY */
    top: auto;
    right: auto;
    bottom: auto;
    left: auto;

    img {
      max-width: 100%;
    }

    &::after {
      width: 0;
      height: 0;
      border-style: solid;
      content: " ";
      display: block;
      position: absolute;

      ${tooltipArrowStyle};
      ${resolveTooltipArrowPosition};
    }

    ${media.largeMobile(css`
      width: auto;
    `)};

    ${resolveTooltipPosition};
`}
`;

// $FlowFixMe: https://github.com/flow-typed/flow-typed/issues/3653#issuecomment-568539198
StyledFormFeedbackTooltip.defaultProps = {
  theme: defaultTheme,
};

const StyledTooltipContent = styled.div`
  ${({ theme }) => css`
    font-family: ${theme.orbit.fontFamily};
    font-size: ${theme.orbit.fontSizeTextNormal};
    font-weight: ${theme.orbit.fontWeightNormal};
    line-height: ${theme.orbit.lineHeightText};
    color: ${theme.orbit.paletteWhite};

    & ${StyledText}, ${Item}, a {
      color: ${theme.orbit.paletteWhite};
      font-weight: ${theme.orbit.fontWeightNormal};
      color: ${theme.orbit.paletteInkNormal};
    }

    ${media.largeMobile(css`
      font-size: ${theme.orbit.fontSizeTextSmall};
      font-weight: ${theme.orbit.fontWeightMedium};

      & ${StyledText}, ${Item}, a {
        color: ${theme.orbit.paletteWhite};
        font-weight: ${theme.orbit.fontWeightMedium};
        font-size: ${theme.orbit.fontSizeTextSmall};
      }
      & a:hover {
        color: ${theme.orbit.paletteWhite};
      }
    `)};
  `}
`;

// $FlowFixMe: https://github.com/flow-typed/flow-typed/issues/3653#issuecomment-568539198
StyledTooltipContent.defaultProps = {
  theme: defaultTheme,
};

const StyledCloseButton = styled.a`
  ${({ theme }) => css`
    color: ${theme.orbit.paletteWhite};
    cursor: pointer;
    margin: ${rtlSpacing(`0 0 0 ${theme.orbit.spaceSmall}`)};
    display: flex;
  `}
`;

// $FlowFixMe: https://github.com/flow-typed/flow-typed/issues/3653#issuecomment-568539198
StyledCloseButton.defaultProps = {
  theme: defaultTheme,
};

const FormFeedbackTooltip = ({
  boundingRef,
  inputSize,
  iconBoundingRef,
  shown = true,
  children,
  isHelp = false,
  inlineLabel,
}: // onClick,
Props): React.Node => {
  const contentRef = React.useRef(null);
  const dimensions = useDimensions(
    { boundingRef, contentRef, iconBoundingRef },
    children,
    inlineLabel,
  );

  const preferredPosition = React.useMemo(
    () =>
      dimensions.bounding.top - dimensions.contentBounding.height > 0
        ? POSITIONS.TOP
        : POSITIONS.BOTTOM,
    [dimensions.bounding.top, dimensions.contentBounding.height],
  );

  return (
    <StyledFormFeedbackTooltip
      ref={contentRef}
      inputSize={inputSize}
      contentBounding={dimensions.contentBounding}
      bounding={dimensions.bounding}
      iconBounding={dimensions.iconBounding}
      position={preferredPosition}
      shown={shown && dimensions.set}
      isHelp={isHelp}
      inlineLabel={inlineLabel}
      aria-live="polite"
    >
      <StyledTooltipContent>{children}</StyledTooltipContent>
      {/* {isHelp && (
        <StyledCloseButton
          tabIndex={0}
          // $FlowFixMe: TODO
          onKeyDown={handleKeyDown(onClick)}
          onClick={ev => {
            ev.preventDefault();
            if (onClick) onClick(ev);
          }}
        >
          <CloseIc size="small" />
        </StyledCloseButton>
      )} */}
    </StyledFormFeedbackTooltip>
  );
};

export default FormFeedbackTooltip;
