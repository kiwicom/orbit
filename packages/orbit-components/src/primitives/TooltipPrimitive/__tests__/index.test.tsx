import * as React from "react";
import { render } from "@testing-library/react";

import Tooltip from "..";
import Airplane from "../../../icons/Airplane";

describe("Tooltip", () => {
  it("it should match snapshot", () => {
    const content = "Write some message to the user";
    const { container } = render(
      <Tooltip content={content}>
        <Airplane />
      </Tooltip>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
