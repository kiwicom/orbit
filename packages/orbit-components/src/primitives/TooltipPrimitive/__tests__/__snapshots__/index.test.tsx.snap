// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip it should match snapshot 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  max-width: 100%;
}

.c0:focus:active {
  outline: none;
}

.c0 [disabled] {
  pointer-events: none;
}

<span
  aria-describedby="1"
  class="c0"
  tabindex="0"
>
  <p>
    Some text
  </p>
</span>
`;
